File: code/06-search-trees/java/AVLTree.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** lecture-notes/code/lecture06/java/AVLTree.java와 동일(verbatim).
 * Distinct int keys. Empty height=-1, leaf height=0, BF=left-right. */
public final class AVLTree {
    // snippet:avl-tree:start
    private static final class Node {int key,height;Node left,right;Node(int k){key=k;}}
    private Node root;private int size;
    private static int h(Node x){return x==null?-1:x.height;}
    private static void update(Node x){x.height=1+Math.max(h(x.left),h(x.right));}
    private static int bf(Node x){return h(x.left)-h(x.right);}
    public boolean contains(int k){Node x=root;while(x!=null){if(k==x.key)return true;x=k<x.key?x.left:x.right;}return false;}
    public boolean insert(int k){if(contains(k))return false;root=insert(root,k);size++;return true;}
    private static Node insert(Node x,int k){if(x==null)return new Node(k);if(k<x.key)x.left=insert(x.left,k);else x.right=insert(x.right,k);return rebalance(x);}
    private static Node rebalance(Node x){update(x);if(bf(x)>1){if(bf(x.left)<0)x.left=left(x.left);return right(x);}if(bf(x)<-1){if(bf(x.right)>0)x.right=right(x.right);return left(x);}return x;}
    private static Node right(Node y){Node x=y.left,b=x.right;x.right=y;y.left=b;update(y);update(x);return x;}
    private static Node left(Node x){Node y=x.right,b=y.left;y.left=x;x.right=b;update(x);update(y);return y;}
    public List<Integer> inorder(){List<Integer>o=new ArrayList<>();inorder(root,o);return List.copyOf(o);}
    private static void inorder(Node x,List<Integer>o){if(x!=null){inorder(x.left,o);o.add(x.key);inorder(x.right,o);}}
    public int height(){return h(root);}
    // snippet:avl-tree:end
    public boolean delete(int k){if(!contains(k))return false;root=delete(root,k);size--;return true;}
    private static Node delete(Node x,int k){
        if(k<x.key)x.left=delete(x.left,k);else if(k>x.key)x.right=delete(x.right,k);
        else{if(x.left==null)return x.right;if(x.right==null)return x.left;Node y=x.right;while(y.left!=null)y=y.left;x.key=y.key;x.right=delete(x.right,y.key);}
        return rebalance(x);
    }

    private static String join(List<Integer> a) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < a.size(); i++) { if (i > 0) sb.append(','); sb.append(a.get(i)); }
        return sb.toString();
    }

    public static void main(String[] args) {
        Map<String, int[]> cases = new LinkedHashMap<>();
        cases.put("LL", new int[] {30, 20, 10});
        cases.put("RR", new int[] {10, 20, 30});
        cases.put("LR", new int[] {30, 10, 20});
        cases.put("RL", new int[] {10, 30, 20});
        for (Map.Entry<String, int[]> e : cases.entrySet()) {
            AVLTree t = new AVLTree();
            for (int k : e.getValue()) t.insert(k);
            System.out.println(e.getKey() + " inorder: " + join(t.inorder()));
            System.out.println(e.getKey() + " height: " + t.height());
        }
        AVLTree big = new AVLTree();
        for (int i = 0; i < 15; i++) big.insert(i);
        System.out.println("sequential15 height: " + big.height());
        System.out.println("sequential15 inorder: " + join(big.inorder()));
    }
}

File: code/06-search-trees/python/avl_tree.py
"""AVLTree: distinct int key. empty height=-1, leaf height=0, BF=left-right.
lecture-notes/code/lecture06/java/AVLTree.java와 같은 정책·예제 입력을 쓴다."""


# snippet:avl-tree:start
class _Node:
    def __init__(self, key):
        self.key = key
        self.height = 0
        self.left = self.right = None


def _h(x):
    return -1 if x is None else x.height


def _update(x):
    x.height = 1 + max(_h(x.left), _h(x.right))


def _bf(x):
    return _h(x.left) - _h(x.right)


def _rotate_right(y):
    x = y.left
    b = x.right
    x.right = y
    y.left = b
    _update(y)
    _update(x)
    return x


def _rotate_left(x):
    y = x.right
    b = y.left
    y.left = x
    x.right = b
    _update(x)
    _update(y)
    return y


def _rebalance(x):
    _update(x)
    if _bf(x) > 1:
        if _bf(x.left) < 0:
            x.left = _rotate_left(x.left)
        return _rotate_right(x)
    if _bf(x) < -1:
        if _bf(x.right) > 0:
            x.right = _rotate_right(x.right)
        return _rotate_left(x)
    return x


class AVLTree:
    def __init__(self):
        self.root = None
        self._size = 0

    def __len__(self):
        return self._size

    def contains(self, key):
        x = self.root
        while x is not None:
            if key == x.key:
                return True
            x = x.left if key < x.key else x.right
        return False

    def insert(self, key):
        if self.contains(key):
            return False
        self.root = self._insert(self.root, key)
        self._size += 1
        return True

    def _insert(self, x, key):
        if x is None:
            return _Node(key)
        if key < x.key:
            x.left = self._insert(x.left, key)
        else:
            x.right = self._insert(x.right, key)
        return _rebalance(x)

    def height(self):
        return _h(self.root)

    def inorder(self):
        out = []

        def walk(x):
            if x is not None:
                walk(x.left)
                out.append(x.key)
                walk(x.right)

        walk(self.root)
        return out
# snippet:avl-tree:end

    def delete(self, key):
        if not self.contains(key):
            return False
        self.root = self._delete(self.root, key)
        self._size -= 1
        return True

    def _delete(self, x, key):
        if key < x.key:
            x.left = self._delete(x.left, key)
        elif key > x.key:
            x.right = self._delete(x.right, key)
        else:
            if x.left is None:
                return x.right
            if x.right is None:
                return x.left
            y = x.right
            while y.left is not None:
                y = y.left
            x.key = y.key
            x.right = self._delete(x.right, y.key)
        return _rebalance(x)

if __name__ == "__main__":
    cases = {
        "LL": [30, 20, 10],
        "RR": [10, 20, 30],
        "LR": [30, 10, 20],
        "RL": [10, 30, 20],
    }
    for name, seq in cases.items():
        t = AVLTree()
        for k in seq:
            t.insert(k)
        print(f"{name} inorder:", ",".join(str(k) for k in t.inorder()))
        print(f"{name} height:", t.height())

    big = AVLTree()
    for i in range(15):
        big.insert(i)
    print("sequential15 height:", big.height())
    print("sequential15 inorder:", ",".join(str(k) for k in big.inorder()))
